### PR TITLE
Release SOP script

### PR DESCRIPTION
**Pull request by dskloetd**
# Motivation

Following the release SOP [checklist](https://www.notion.so/dfinityorg/Checklist-YYYY-MM-DD-0e1db3aadfc140ab9bd9c2f3957ba25d) is a lot of manual work.
A script should help with that.
This PR adds a script that helps with the stops up to the point where there RC is on staging and a post is made on Slack.
The plan is to add more steps in follow-up PRs.

I wanted to achieve the following goals with the script:
* Every command that modifies state should be confirmed by the user.
* The script should record the same data (such as commit, WASM hash, testnet URL) as we record in the checklist.
* If the script fails or is aborted halfway through, it should be possible to continue where it left off.
* When the script continues where it left off, it verifies that the previously recorded state is still correct.

To keep track of progress, the script creates a JSON file based on the release name. Each step is recorded in the JSON file and if the script is rerun later, previously recorded steps are skipped.

The format of the JSON file is like this:
```
[
  {
    "name": "Start time",
    "value": "2023-06-29 16:17:36"
  },
  {
    "name": "git fetch --tags --force",
    "value": "Done"
  },
  {
    "name": "git checkout -b Release-2023-06-29 origin/main",
    "value": "Done"
  },
  {
    "name": "Release branch",
    "value": "Release-2023-06-29"
  },
  {
    "name": "Current branch",
    "value": "Release-2023-06-29"
  },
  {
    "name": "git push origin Release-2023-06-29",
    "value": "Done"
  },
  {
    "name": "Release branch URL",
    "value": "https://github.com/dfinity/nns-dapp/tree/Release-2023-06-29"
  },
  {
    "name": "git tag -f release-candidate",
    "value": "Done"
  },
  {
    "name": "Release commit",
    "value": "85710a82e0c4b287b43a7c97a8e6b2f231f800e4"
  },
  {
    "name": "tags/release-candidate commit",
    "value": "85710a82e0c4b287b43a7c97a8e6b2f231f800e4"
  },
  {
    "name": "git push origin -f tags/release-candidate",
    "value": "Done"
  },
  {
    "name": "Security test: npm audit",
    "value": "https://github.com/advisories/GHSA-j8xg-fqg3-53r7"
  },
  {
    "name": "Expected npm audit",
    "value": "https://github.com/advisories/GHSA-j8xg-fqg3-53r7"
  },
  {
    "name": "Security test: cargo audit",
    "value": "RUSTSEC-2020-0071,RUSTSEC-2020-0159,RUSTSEC-2023-0044"
  },
  {
    "name": "Expected cargo audit",
    "value": "RUSTSEC-2020-0071,RUSTSEC-2020-0159,RUSTSEC-2023-0044"
  },
  {
    "name": "Security test: no HTML in templates",
    "value": "Checked"
  },
  {
    "name": "scripts/nns-dapp/download-ci-wasm --commit tags/release-candidate --dir /Users/dskloet/dev/nns-dapp/tree2/release/ci --wasm-filename nns-dapp.wasm.gz",
    "value": "Done"
  },
  {
    "name": "CI NNS-dapp WASM hash",
    "value": "d3f2a0e4f77b6b6321195dfac674a5abd22697be0f4cd812df00db296f01c4b9"
  },
  {
    "name": "Local WASM hash",
    "value": "d3f2a0e4f77b6b6321195dfac674a5abd22697be0f4cd812df00db296f01c4b9"
  },
  {
    "name": "scripts/canister_ids2 --remove --network staging --canister nns-dapp",
    "value": "Done"
  },
  {
    "name": "dfx identity",
    "value": "default: s2dw5-fmktb-knrnq-772tr-ca3g7-me224-st35j-f3eli-xx2vf-aoesd-2qe"
  },
  {
    "name": "dfx canister create nns-dapp --network staging",
    "value": "Done"
  },
  {
    "name": "NNS-dapp staging canister ID",
    "value": "jrc6w-ayaaa-aaaaa-aac3a-cai"
  },
  {
    "name": "Current NNS-dapp staging canister ID",
    "value": "jrc6w-ayaaa-aaaaa-aac3a-cai"
  },
  {
    "name": "DFX_NETWORK=staging ./config.sh",
    "value": "Done"
  },
  {
    "name": "dfx canister install nns-dapp --argument $'(opt record{\\n  args = vec {\\n    record{ 0=\"CKBTC_INDEX_CANISTER_ID\"; 1=\"si2b5-pyaaa-aaaaa-aaaja-cai\" };\\n    record{ 0=\"CKBTC_LEDGER_CANISTER_ID\"; 1=\"q3fc5-haaaa-aaaaa-aaahq-cai\" };\\n    record{ 0=\"CKBTC_MINTER_CANISTER_ID\"; 1=\"sgymv-uiaaa-aaaaa-aaaia-cai\" };\\n    record{ 0=\"CYCLES_MINTING_CANISTER_ID\"; 1=\"rkp4c-7iaaa-aaaaa-aaaca-cai\" };\\n    record{ 0=\"DFX_NETWORK\"; 1=\"staging\" };\\n    record{ 0=\"FEATURE_FLAGS\"; 1=\"{\\\\\"ENABLE_CKBTC\\\\\":true,\\\\\"ENABLE_CKTESTBTC\\\\\":false,\\\\\"ENABLE_INSTANT_UNLOCK\\\\\":false,\\\\\"ENABLE_SIMULATE_MERGE_NEURONS\\\\\":false,\\\\\"ENABLE_SNS_AGGREGATOR\\\\\":true,\\\\\"ENABLE_SNS_VOTING\\\\\":true}\" };\\n    record{ 0=\"FETCH_ROOT_KEY\"; 1=\"true\" };\\n    record{ 0=\"GOVERNANCE_CANISTER_ID\"; 1=\"rrkah-fqaaa-aaaaa-aaaaq-cai\" };\\n    record{ 0=\"HOST\"; 1=\"https://nnsdapp.dfinity.network\" };\\n    record{ 0=\"IDENTITY_SERVICE_URL\"; 1=\"https://qhbym-qaaaa-aaaaa-aaafq-cai.nnsdapp.dfinity.network\" };\\n    record{ 0=\"LEDGER_CANISTER_ID\"; 1=\"ryjl3-tyaaa-aaaaa-aaaba-cai\" };\\n    record{ 0=\"OWN_CANISTER_ID\"; 1=\"jrc6w-ayaaa-aaaaa-aac3a-cai\" };\\n    record{ 0=\"OWN_CANISTER_URL\"; 1=\"https://jrc6w-ayaaa-aaaaa-aac3a-cai.nnsdapp.dfinity.network\" };\\n    record{ 0=\"ROBOTS\"; 1=\"<meta name=\\\\\"robots\\\\\" content=\\\\\"noindex, nofollow\\\\\" />\" };\\n    record{ 0=\"SNS_AGGREGATOR_URL\"; 1=\"https://lzybt-syaaa-aaaaa-aacxa-cai.nnsdapp.dfinity.network\" };\\n    record{ 0=\"TVL_CANISTER_ID\"; 1=\"\" };\\n    record{ 0=\"WASM_CANISTER_ID\"; 1=\"qaa6y-5yaaa-aaaaa-aaafa-cai\" };\\n  };\\n})' --network staging --wasm ./release/ci/nns-dapp.wasm.gz",
    "value": "Done"
  },
  {
    "name": "Staging URL",
    "value": "https://jrc6w-ayaaa-aaaaa-aac3a-cai.nnsdapp.dfinity.network/"
  },
  {
    "name": "Staging WASM hash",
    "value": "d3f2a0e4f77b6b6321195dfac674a5abd22697be0f4cd812df00db296f01c4b9"
  },
  {
    "name": "Post on Slack",
    "value": "https://dfinity.slack.com/archives/C03HZERHBT3/p1688048643485249"
  }
]
```

There are 3 kind of steps in the checklist:
1. Performing an action
2. Recording some data
3. Verifying that some values match

To reflect this, the script has 3 functions:
`confirm_cmd`: This runs a given command (only if it wasn't run already) after confirmation from the user and records in the JSON file that the command was run. Example: Download the WASM from CI.
`record`: This runs a command and records the result in the JSON file. The command is assumed to not change state but only read state so no confirmation is requested from the user. Example: get the WASM hash from CI.
`verify`: Runs a command and verifies that the output matches a recorded value. Example: Get the hash of the downloaded WASM and compare with the hash from CI.

You can also pass a JSON file specifying mock behavior. This is used to test several scenarios.

# Changes

1. Add the `release-sop` script.
2. Add a test.
3. Add the JSON files with the mocks behavior and test expectations.

# Tests

Unit test is added and the last 2 RCs (2023-06-28-v2 and 2023-06-29) were created with the script.

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

File: scripts/nns-dapp/release-sop [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+	checked out and run:
+	  ./release-sop.sh --continue
+
+	EOF
+}
+
+# Source the clap.bash file ---------------------------------------------------
+source "$SOURCE_DIR/../clap.bash"
+# Define options
+clap.define short=n long=new desc="Start a new release branch" variable=IS_NEW_RELEASE nargs=0 default="false"
```
Comment (mstrasinskis) [inline scripts/nns-dapp/release-sop:27]: Currently we are using `staging` for the tests. But it is not the rule and sometimes we switch to different testnets when staging is outdated. Maybe it's not relevant anymore. Does it make sense to reserve the `-n` flag for the network (as in some other scripts) and use different flag to start a new release?

File: scripts/nns-dapp/release-sop [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+confirm_cmd git tag -f release-candidate
+record "Release commit" git rev-parse release-candidate
+verify "Release commit" "tags/release-candidate commit" git rev-parse release-candidate
+confirm_cmd git push origin -f tags/release-candidate
+record "Security test: npm audit" npm_audit
+verify "Security test: npm audit" "Expected npm audit" echo "https://github.com/advisories/GHSA-j8xg-fqg3-53r7"
+record "Security test: cargo audit" cargo_audit
+verify "Security test: cargo audit" "Expected cargo audit" echo "RUSTSEC-2020-0071,RUSTSEC-2020-0159,RUSTSEC-2023-0044"
+record "Security test: no HTML in templates" no_html_in_templates
+confirm_cmd "$SOURCE_DIR/download-ci-wasm" --commit tags/release-candidate --dir "$TOP_DIR/release/ci" --wasm-filename nns-dapp.wasm.gz
```
Comment (mstrasinskis) [inline scripts/nns-dapp/release-sop:344]: Should the checklist also be updated to use the wasm filename w/ extention dapp.wasm`.gz`?

**Review by mstrasinskis — APPROVED** [review]
Great, thanks! Just a couple of minor comments.

File: scripts/nns-dapp/release-sop (hunk 1 above)
Comment (dskloetd) [reply scripts/nns-dapp/release-sop]: The short flag names weren't optional before but now they are so I've removed all the short flag names for now.
So if we do add a network flag later, we can use `-n` for it.

**Pull request merged**